src/app-layer-ftp.c

```diff
@@ -414,15 +532,217 @@ static int FTPRegisterPatternsForProtocolDetection(void)
     return 0;
 }
 
+
+static StreamingBufferConfig sbcfg = STREAMING_BUFFER_CONFIG_INITIALIZER;
+
+/**
+ * \brief This function is called to retrieve a ftp request
+ * \param ftp_state the ftp state structure for the parser
+ * \param input input line of the command
+ * \param input_len length of the request
+ * \param output the resulting output
+ *
+ * \retval 1 when the command is parsed, 0 otherwise
+ */
+static int FTPDataParse(Flow *f, void *ftp_state,
+        AppLayerParserState *pstate,
+        uint8_t *input, uint32_t input_len,
+        void *local_data, int direction)
+{
+    FtpDataState *ftpdata_state = (FtpDataState *)ftp_state;
+    uint16_t flags = FileFlowToFlags(f, direction);
+    int ret = 0;
+    /* we depend on detection engine for file pruning */
+    flags |= FILE_USE_DETECT;
+    if (ftpdata_state->files == NULL) {
+        const char *data = (char *)FlowGetStorageById(f, AppLayerExpectationGetDataId());
+        const char *filename;
+        if (data == NULL) {
+            SCReturnInt(-1);
+        }
+
+        ftpdata_state->files = FileContainerAlloc();
+        if (ftpdata_state->files == NULL) {
+            SCLogError(SC_ERR_MEM_ALLOC, "Could not create file container");
+            SCReturnInt(-1);
```

Comment:
should FlowFreeStorageById(f, AppLayerExpectationGetDataId()); be called in this case as well?